### PR TITLE
Purge addresses that fail to reach a peer

### PR DESCRIPTION
**Pull request by tomaka**
So far, the `add_known_address` method of the networking was meant to be used to inject addresses that are permanent. In other words, addresses for reserved nodes or bootnodes that the user has explicitly configured.

However, Polkadot now uses this method to add the addresses reported by the authority-discovery system. In other words, addresses reported by the nodes themselves.
Before this PR, these addresses would never be purged. After this PR, addresses that fail to connect to someone will be removed and not tried again until they are re-added.

cc #8466

This PR is sub-optimal, so I don't think we should be closing #8466. I'm doing this change to try to quickly improve the situation with parachains.
Ideally, we would have a proper story for addresses management, including properly documenting the behaviour, and removing addresses for peers whose reputation drops to 0, but these are way bigger changes.

### CODE REVIEW

File: client/network/src/discovery.rs [hunk 1]
```diff
@@ -287,12 +291,14 @@ impl DiscoveryBehaviour {
 	///
 	/// If we didn't know this address before, also generates a `Discovered` event.
 	pub fn add_known_address(&mut self, peer_id: PeerId, addr: Multiaddr) {
-		if self.user_defined.iter().all(|(p, a)| *p != peer_id && *a != addr) {
+		let addrs_list = self.ephemeral_addresses.entry(peer_id).or_default();
+		if !addrs_list.iter().any(|a| *a == addr) {
```
Comment (kpp) [inline client/network/src/discovery.rs:295]: Why you don't check in `permanent_addresses` too?

**Review by kpp — APPROVED** [review]

**Comment by tomaka** [comment]
Yeah, this was just waiting to be merged.

**Comment by tomaka** [comment]
bot merge

File: client/network/src/discovery.rs (hunk 1 above)
Comment (kpp) [reply client/network/src/discovery.rs:295]: @tomaka ping

File: client/network/src/discovery.rs [hunk 2]
```diff
@@ -536,6 +546,12 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 		addr: &Multiaddr,
 		error: &dyn std::error::Error
 	) {
+		if let Some(peer_id) = peer_id {
+			if let Some(list) = self.ephemeral_addresses.get_mut(peer_id) {
+				list.retain(|a| a != addr);
```
Comment (0x7CFE) [inline client/network/src/discovery.rs:551]: Looks like this chunk can be simplified using combinators.

**Review by 0x7CFE — APPROVED** [review]
LGTM

File: client/network/src/discovery.rs (hunk 2 above)
Comment (gilescope) [reply client/network/src/discovery.rs:551]: It's pretty readable as is.

**Comment by kpp** [comment]
Yes

**Comment by bkchr** [comment]
@kpp could you merge master so that we can merge this pr?

**Comment by kpp** [comment]
Yes. In an hour

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
ty @kpp

**Comment by bkchr** [comment]
bot merge

**Pull request merged**